x/poolmanager/protorev.go

```diff
@@ -1,91 +1,126 @@
 package poolmanager
 
 import (
+	"fmt"
+
 	sdk "github.com/cosmos/cosmos-sdk/types"
 	gogotypes "github.com/cosmos/gogoproto/types"
 
+	"github.com/osmosis-labs/osmosis/osmomath"
 	"github.com/osmosis-labs/osmosis/osmoutils"
 	"github.com/osmosis-labs/osmosis/v21/x/poolmanager/types"
+
+	"github.com/cosmos/cosmos-sdk/store/prefix"
 )
 
-// IncreaseTakerFeeTrackerForStakers gets the current value of the taker fee tracker for stakers, adds the given amount to it, and sets the new value.
-func (k Keeper) IncreaseTakerFeeTrackerForStakers(ctx sdk.Context, takerFeeForStakers sdk.Coin) {
-	currentTakerFeeForStakers := k.GetTakerFeeTrackerForStakers(ctx)
-	if !takerFeeForStakers.IsZero() {
-		newTakerFeeForStakersCoins := currentTakerFeeForStakers.Add(takerFeeForStakers)
-		newTakerFeeForStakers := types.TrackedVolume{
-			Amount: newTakerFeeForStakersCoins,
+// GetTakerFeeTrackerForStakers returns the taker fee for stakers tracker for all denoms that has been
+// collected since the accounting height.
+func (k Keeper) GetTakerFeeTrackerForStakers(ctx sdk.Context) []sdk.Coin {
+	takerFeesForStakers := make([]sdk.Coin, 0)
+
+	store := ctx.KVStore(k.storeKey)
+	iterator := sdk.KVStorePrefixIterator(store, types.KeyTakerFeeStakersProtoRevArray)
+
+	defer iterator.Close()
+	for ; iterator.Valid(); iterator.Next() {
+		bz := iterator.Value()
+		takerFeeForStakers := sdk.Coin{}
+		if err := takerFeeForStakers.Unmarshal(bz); err == nil {
+			takerFeesForStakers = append(takerFeesForStakers, takerFeeForStakers)
 		}
-		osmoutils.MustSet(ctx.KVStore(k.storeKey), types.KeyTakerFeeStakersProtoRev, &newTakerFeeForStakers)
 	}
+
+	return takerFeesForStakers
 }
 
-// IncreaseTakerFeeTrackerForCommunityPool gets the current value of the taker fee tracker for the community pool, adds the given amount to it, and sets the new value.
-func (k Keeper) IncreaseTakerFeeTrackerForCommunityPool(ctx sdk.Context, takerFeeForCommunityPool sdk.Coin) {
-	currentTakerFeeForCommunityPool := k.GetTakerFeeTrackerForCommunityPool(ctx)
-	if !takerFeeForCommunityPool.IsZero() {
-		newTakerFeeForCommunityPoolCoins := currentTakerFeeForCommunityPool.Add(takerFeeForCommunityPool)
-		newTakerFeeForCommunityPool := types.TrackedVolume{
-			Amount: newTakerFeeForCommunityPoolCoins,
-		}
-		osmoutils.MustSet(ctx.KVStore(k.storeKey), types.KeyTakerFeeCommunityPoolProtoRev, &newTakerFeeForCommunityPool)
+// GetTakerFeeTrackerForStakersByDenom returns the taker fee for stakers tracker for the specified denom that has been
+// collected since the accounting height.
+func (k Keeper) GetTakerFeeTrackerForStakersByDenom(ctx sdk.Context, denom string) (sdk.Coin, error) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyTakerFeeStakersProtoRevArray)
+	key := types.GetKeyPrefixTakerFeeStakersProtoRevByDenom(denom)
+
+	bz := store.Get(key)
+	if len(bz) == 0 {
+		return sdk.NewCoin(denom, osmomath.ZeroInt()), fmt.Errorf("no taker fees for stakers for denom %s", denom)
```

Comment:
If this is an error condition, we should be returning `sdk.Coin{}`

If this is not an error, then let's return only zero coin and nil.

Returning both is a code smell that might lead to difficulty for the client dealing with this down the road

Reply:
It is not an error, it is the case when a new token is added to the tracker.

Changed to nil error return here https://github.com/osmosis-labs/osmosis/pull/7240/commits/2e78405133dd0ef3a4f0db5ec0af47607cf2bbff